File: coding/src/com/algos/core/li05_binarysearch/revised/L74_SearchA2DMatrix.java
package com.algos.core.li05_binarysearch.revised;

public class L74_SearchA2DMatrix {

    public static void main(String[] args) {
        L74_SearchA2DMatrix obj = new L74_SearchA2DMatrix();
        int[][] matrix = new int[][] {
                        {1,3,5,7},
                        {10,11,16,20},
                        {23,30,34,60}
                    };

        System.out.println(obj.searchMatrix(matrix, 5));
    }

    // Problem: https://leetcode.com/problems/search-a-2d-matrix/
    // Idea: Each row and columns are in increasing order and first element of next
    // row is greater then last element of curr row.
    // For this case optimal approach is to apply BS on range 0 to n*m-1 as the
    // elments are continueously increasing from 0 to n*m-1.
    public boolean searchMatrix(int[][] mat, int target) {
        int m = mat.length, n = mat[0].length;
        int low = 0, high = m * n - 1;
        while (low <= high) {
            int mid = (low + high) / 2;
            int row = mid / n, col = mid % n;
            if (mat[row][col] == target) {
                return true;
            } else if (target > mat[row][col]) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return false;
    }
}

File: dsa/concepts/05_binarysearch/python/l74_search_a2_dmatrix.py
"""
LeetCode 74 - Search a 2D Matrix

Idea:
Treat the 2D matrix as a sorted 1D array.
Apply binary search on indices [0, m*n - 1].

Mapping:
row = mid // n
col = mid % n

Time Complexity: O(log(m*n))
Space Complexity: O(1)
"""

def search_matrix(matrix, target):
    if not matrix or not matrix[0]:
        return False

    m, n = len(matrix), len(matrix[0])
    low, high = 0, m * n - 1

    while low <= high:
        mid = (low + high) // 2
        row, col = mid // n, mid % n

        if matrix[row][col] == target:
            return True
        elif matrix[row][col] < target:
            low = mid + 1
        else:
            high = mid - 1

    return False


# Driver code
if __name__ == "__main__":
    matrix = [
        [1, 3, 5, 7],
        [10, 11, 16, 20],
        [23, 30, 34, 60]
    ]
    print(search_matrix(matrix, 5))  # True
